server/apps/microservices/src/processors/machine-learning.processor.ts

```diff
@@ -45,24 +45,47 @@ export class MachineLearningProcessor {
     try {
       const { asset }: { asset: AssetEntity } = job.data;
 
-      const res = await axios.post(
-        immich_machine_learning_url + '/object-detection/detect-object',
-        {
-          thumbnailPath: asset.resizePath,
-        },
-      );
+      const res = await axios.post(immich_machine_learning_url + '/object-detection/detect-object', {
+        thumbnailPath: asset.resizePath,
+      });
 
       if (res.status == 201 && res.data.length > 0) {
-        const smartInfo = new SmartInfoEntity();
-        smartInfo.assetId = asset.id;
-        smartInfo.objects = [...res.data];
+        this.createAndSaveTagToAsset(asset, res.data);
+      }
+    } catch (e) {
+      this.logger.error(e);
+    }
+  }
+
+  private async createAndSaveTagToAsset(asset: AssetEntity, tagNames: string[]) {
+    try {
+      const immichTags: TagEntity[] = [];
+
+      for (const tag of tagNames) {
+        let immichTag: TagEntity | null;
+
+        immichTag = await this.tagRepository.findOne({ where: { name: tag, userId: asset.userId } });
+
+        if (!immichTag) {
+          immichTag = new TagEntity();
+          immichTag.name = tag;
+          immichTag.type = TagType.OBJECT;
+          immichTag.userId = asset.userId;
+
+          immichTag = await this.tagRepository.save(immichTag);
+        }
+
+        immichTags.push(immichTag);
+      }
+
+      const assetEntity = await this.assetRepository.findOne({ where: { id: asset.id } });
 
-        await this.smartInfoRepository.upsert(smartInfo, {
-          conflictPaths: ['assetId'],
-        });
+      if (assetEntity) {
+        assetEntity.tags = assetEntity.tags ? Array.from(new Set([...assetEntity.tags, ...immichTags])) : immichTags;
+        await this.assetRepository.save(assetEntity);
```

Comment:
Maybe this would be cleaner as `assetRepository.addTags(assetEntity, tags)`? I'm not sure what the idioms in nest/typeorm are though.

Reply:
I believe save with cascades would do this automatically if it could "match" on primary key. So, if primary key was userId/type/name it would upsert with a simple `.save()`. The problem is, at the table level, the only thing required to be unique is id, so typeorm would just create new records since the tags don't have ids yet. Which is why we're having to add code here to manually check if the tag exists already or not.

Reply:
I also don't think this work the way you think it works. This will de-duplicate an array, but using object equality (==), and I don't think these object will ever be "equal" in that sense. I think what you want to do is de-duplicate based on id or userId/name, which this isn't doing.